sympy/core/tests/test_evalf.py

```diff
@@ -317,11 +317,11 @@ def test_evalf_power_subs_bugs():
     assert (x**2).evalf(subs={x: 0}) == 0
     assert sqrt(x).evalf(subs={x: 0}) == 0
     assert (x**Rational(2, 3)).evalf(subs={x: 0}) == 0
-    assert (x**x).evalf(subs={x: 0}) == 1
-    assert (3**x).evalf(subs={x: 0}) == 1
-    assert exp(x).evalf(subs={x: 0}) == 1
-    assert ((2 + I)**x).evalf(subs={x: 0}) == 1
-    assert (0**x).evalf(subs={x: 0}) == 1
+    assert (x**x).evalf(subs={x: 0}) == 1.0
+    assert (3**x).evalf(subs={x: 0}) == 1.0
+    assert exp(x).evalf(subs={x: 0}) == 1.0
+    assert ((2 + I)**x).evalf(subs={x: 0}) == 1.0
+    assert (0**x).evalf(subs={x: 0}) == 1.0
```

Comment:
Would it be possible (or sensible) to lessen the pain so that `Float(1.0) == 1` even thought `Float(1.0) != Rational(1)`?

Reply:
Does that lessen the pain? I think it would add to the confusion.

I've tried to preserve the behaviour that if `x` is a sympy object then `x == y` iff `x == sympify(y)`. Implicit sympification should make it so that Float/float are interchangeable and also int/Integer are interchangeable provided that some sympy type is involved somewhere.